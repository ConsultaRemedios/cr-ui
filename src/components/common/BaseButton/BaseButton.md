### Sizes

```js
  <div>
    <BaseButton size="small">Button small</BaseButton>
    <BaseButton size="medium">Button medium</BaseButton>
    <BaseButton size="large">Button large</BaseButton>
  </div>
```

### Types

```js 
  <div>
    <BaseButton type="success">Button success</BaseButton>
    <BaseButton type="info">Button info</BaseButton>
    <BaseButton type="warning">Button warning</BaseButton>
    <BaseButton type="neutral">Button neutral</BaseButton>
    <BaseButton type="danger">Button danger</BaseButton>
    <BaseButton type="naked">Button naked</BaseButton>
    <BaseButton type="naked-red">Button naked-red</BaseButton>
  </div>
```

### Fill

Fill property forces the button width to fill 100% width of its parent.

```js
<div>
  <div :style="{ marginBottom: '15px' }">
    <BaseButton :fill="true">Example 1</BaseButton>
  </div>
  <div>
    <BaseButton type="neutral" :fill="true">Example 2</BaseButton>
  </div>
</div>
```

### Combining type and size props

```js
  <div>
    <BaseButton type="neutral" size="large">Button neutral large</BaseButton>
    <BaseButton type="success" size="medium">Button success medium</BaseButton>
  </div>
```

### Button as an achor tag

Just add a `path` prop with the url you want to link to as the value.
You can also open the link in a new tab using the `openNewTab` prop as true

```js
  <div>
    <BaseButton type="neutral" path="http://www.google.com">
      Anchor default
    </BaseButton>

    <BaseButton type="success" size="large" path="http://www.google.com" :openNewTab="true">
      Anchor new tab
    </BaseButton>
  </div>
```

### Disabled State

```js
<div>
  <BaseButton :disabled="true">Button neutral</BaseButton>
  <BaseButton :disabled="true" type="naked">Button naked</BaseButton>
  <BaseButton :disabled="true" type="naked-red">Button naked red</BaseButton>
</div>
```

### Inverted option

Invert the background-color and border-color by adding this option

```js
<div>
  <BaseButton :inverted="true" type="success">Success inverted</BaseButton>
  <BaseButton :inverted="true" type="info">Info inverted</BaseButton>
  <BaseButton :inverted="true" type="warning">Warning inverted</BaseButton>
  <BaseButton :inverted="true" type="danger">Danger inverted</BaseButton>
  <BaseButton :inverted="true" type="neutral">Neutral inverted</BaseButton>
</div>
```

### Loading button

Add a behavior of loading to button

```js
const loading = false;

<div>
  <label>
    is Loading:
    <input type="checkbox" @input="loading = !loading"/>
  </label>

  <br/>
  <br/>

  <BaseButton type="success" :isLoading="loading">Loading button</BaseButton>
  <BaseButton type="neutral" :isLoading="loading">Loading button</BaseButton>
</div>
```


### Button with icon

Add a icon in button

```js
  const getIcon = (icons, name) => icons.find((icon) => icon.id === `${name}.icon`);

  <div>
      <BaseButton size="small" :icon="getIcon(icons, 'info')" type="success">Button success</BaseButton>
      <BaseButton size="medium" :icon="getIcon(icons, 'cancel')" type="neutral">Button success</BaseButton>
      <BaseButton size="large" :icon="getIcon(icons, 'change')" type="warning">Button success</BaseButton>
  </div>
```