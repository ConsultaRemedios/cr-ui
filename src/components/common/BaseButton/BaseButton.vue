<template>
  <component :is="element"
    :type="buttonType"
    :disabled="disabled || isLoading"
    :class="[
      $style.button,
      $style[size],
      $style[type],
      {
        [$style.inverted]: inverted,
        [$style.disabled]: disabled,
        [$style.fill]: fill,
        [$style.buttonLoading]: isLoading,
        [$style.buttonShadow]: type !== 'naked' && type !== 'naked-red',
      }
    ]"
    :target="target"
    @click="onClick"
    :href="path"
    v-bind="$attrs"
  >
    <BaseIcon
      v-if="isLoading"
      :id="loaderIcon.id"
      :class="$style.loaderIcon"
    />
    <BaseIcon
      v-if="icon"
      :id="icon.id"
      :class="$style.iconButton"
    />
    <span :class="[{[$style.textLoading]: isLoading}]">
      <slot></slot>
    </span>
  </component>
</template>

<script>
  import BaseIcon from '../BaseIcon';
  import loaderIcon from '../../../icons/loader.icon.svg';

  export default {
    inheritAttrs: false,
    name: 'BaseButton',
    props: {
      buttonType: {
        type: String,
        validator(value) {
          return ['button', 'submit'].includes(value);
        },
      },

      size: {
        type: String,
        default: 'medium',
        validator(value) {
          return ['small', 'medium', 'large'].includes(value);
        },
      },

      type: {
        type: String,
        default: 'success',
        validator(value) {
          return ['success', 'info', 'warning', 'danger', 'neutral', 'naked', 'naked-red'].includes(value);
        },
      },

      disabled: {
        type: Boolean,
        default: false,
      },

      fill: {
        type: Boolean,
        default: false,
      },

      path: {
        type: String,
      },

      openNewTab: {
        type: Boolean,
        default: false,
      },

      inverted: {
        type: Boolean,
        default: false,
      },

      isLoading: {
        type: Boolean,
        default: false,
      },
      icon: {
        type: Object,
      },
    },

    computed: {
      element() {
        return this.path ? 'a' : 'button';
      },

      target() {
        return this.openNewTab ? '_blank' : null;
      },

      loaderIcon() {
        return loaderIcon;
      },
    },

    methods: {
      onClick() {
        this.$emit('click');
      },
    },

    components: {
      BaseIcon,
    },
  };
</script>

<style module>
  @value media-md from './../../../styles/variables.css';

  .button > span {
    display: inline-block;
    line-height: 24px;
  }

  .button {
    border-radius: 4px;
    cursor: pointer;
    text-decoration: none;
    display: inline-block;
    text-align: center;
    transition: background-color 0.25s ease-out,
                border-color 0.25s ease-out,
                color 0.25s ease-out;
  }

  .buttonShadow {
    box-shadow: 0 1px 6px 0 rgba(0,0,0,0.15);
  }

  .button:hover {
    text-decoration: none;
  }

  .button:focus {
    outline: none;
  }

  /* Sizes */
  .small {
    font-size: 14px;
    font-weight: 600;
    padding: 1px 8px;
  }

  .medium {
    padding: 7px 18px;
    font-size: 14px;
    font-weight: 600;
  }

  .large {
    font-size: 14px;
    font-weight: 600;
    padding: 10px 20px;
  }

  /* Fill */
  .fill {
    width: 100%;
  }

  /* Types */
  .success {
    background-color: #1BB738;
    border: 1px solid #1BB738;
    color: #FFF;
  }

  .success:focus,
  .success:hover {
    background-color: #16942D;
    border-color: #16942D;
    color: #FFF;
  }

  .info {
    background-color: #3DC2C2;
    border-color: #3DC2C2;
    border-style: solid;
    color: #FFF;
  }

  .info:focus,
  .info:hover {
    background-color: #169998;
    border-color: #169998;
    color: #FFF;
  }

  .warning {
    background-color: #F9AD3D;
    border-color: #F9AD3D;
    border-style: solid;
    color: #FFF;
  }

  .warning:focus,
  .warning:hover {
    background-color: #CA8C31;
    border-color: #CA8C31;
    color: #FFF;
  }

  .neutral {
    background-color: #FFF;
    border: 1px solid #D9D9D9;
    color: #333333;
  }

  .neutral:focus,
  .neutral:hover {
    background-color: #F8F8F8;
    border-color: #D9D9D9;
    color: #333333;
  }

  .danger {
    background-color: #D30B3F;
    border-color: #D30B3F;
    border-style: solid;
    color: #FFF;
  }

  .danger:focus,
  .danger:hover {
    background-color: #AB0933;
    border-color: #AB0933;
    color: #FFF;
  }

  .naked {
    border: 0;
    background: transparent;
    color: #00AAE5;
  }

  .naked:focus,
  .naked:hover {
    color: #007CA7;
  }

  .naked-red {
    composes: naked;
    color: #D30B3F;
  }

  .naked-red:hover {
    color: #AB0933;
  }

  /* States */
  .disabled,
  .disabled:hover,
  .disabled:focus {
    cursor: default;
    background-color: #EEE;
    border-color: #EEE;
    color: #CCC;
  }

  .inverted.success {
    border: 1px solid #1BB738;
    color: #1BB738;
  }

  .inverted.info {
    border: 1px solid #3DC2C2;
    color: #3DC2C2;
  }

  .inverted.warning {
    border: 1px solid #F9AD3D;
    color: #F9AD3D;
  }

  .inverted.neutral {
    background-color: #D9D9D9;
    border: 1px solid #D9D9D9;
  }

  .inverted.neutral:hover {
    background-color: #333;
    border-color: #333;
  }

  .inverted.danger {
    border: 1px solid #D30B3F;
    color: #D30B3F;
  }

  .inverted {
    background-color: #FFF;
  }

  .inverted:hover {
    color: #FFF;
  }

  /* Combined type and size */
  .naked.medium,
  .naked-red.medium {
    font-size: 14px;
    font-weight: 800;
  }

  .naked.disabled,
  .naked.disabled:hover {
    background-color: transparent;
  }

  .loaderIcon {
    animation: rotate-loader 1.5s infinite;

    position:absolute;
    top: 50%;
    left:0;
    right:0;
    margin-left:auto;
    margin-right:auto;
    transform: translateY(-50%);
  }

  .iconButton {
    margin-right: 4px;
  }

  .buttonLoading {
    position: relative;
  }

  .textLoading {
    opacity: 0;
  }

  @media media-md {
    .large {
      padding-left: 25px;
      padding-right: 25px;
    }
  }

  @keyframes rotate-loader {
    0% {
      transform: translateY(-50%) rotate(0deg);
    }

    100% {
      transform: translateY(-50%) rotate(360deg);
    }
  }
</style>
