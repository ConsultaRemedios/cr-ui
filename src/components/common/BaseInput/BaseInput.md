### Basic usage
By default, input fill the parent element size
```js
  <BaseInput placeholder="Type something here..."/>
```

### With error
```js
  <BaseInput :has-error="true"/>
```

### Loading State
```js
  <BaseInput :is-loading="true"/>
```

### Disabled State
```js
  <BaseInput :disabled="true"/>
```
### With prefix or suffix icon
```js
  <BaseInput 
    :prefix-icon="{ icon: $getIcon(allIcons, 'place'), color: '#00AAE5' }" 
    :suffix-icon="{ icon: $getIcon(allIcons, 'search'), color: '#666666' }"
    size="small" 
  />
```