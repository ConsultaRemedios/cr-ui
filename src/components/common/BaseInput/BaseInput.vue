<template>
  <div :class="$style.container">
    <BaseIcon
      v-if="prefixIcon"
      :id="prefixIcon.icon.id"
      :class="[$style.inputIcon, $style.prefixIcon]"
      :style="{ color: prefixIcon.color }"
    />

    <input
      :class="inputClasses"
      :name="name"
      :style="customStyle"
      v-bind="$attrs"
      @keydown="onKeydown"
      @input="onChange"
      @blur="onBlur"
      @focus="onFocus"
    >

    <BaseIcon
      v-if="suffixIcon"
      :id="suffixIcon.icon.id"
      :class="[$style.inputIcon, $style.suffixIcon]"
      :style="{ color: suffixIcon.color }"
    />

    <img
      v-if="isLoading"
      :class="$style.loader"
      src="./../../../assets/loader-input.svg"
    >
  </div>
</template>

<script>
import BaseIcon from '../BaseIcon';

export default {

  name: 'BaseInput',

  components: {
    BaseIcon,
  },
  inheritAttrs: false,

  props: {
    classes: {
      type: Array,
      default: () => [],
    },

    size: {
      type: String,
      default: 'medium',
    },
    name: {
      type: String,
      default: '',
    },

    hasError: {
      type: Boolean,
      default: false,
    },

    isLoading: {
      type: Boolean,
      default: false,
    },

    customStyle: {
      type: Object,
      default: () => {},
    },

    model: {
      type: String,
      default: '',
    },

    prefixIcon: {
      type: Object,
    },

    suffixIcon: {
      type: Object,
    },

  },

  computed: {
    inputClasses() {
      return this.classes.concat([this.$style[this.size], {
        [this.$style.input]: true,
        [this.$style.hasError]: this.hasError,
        [this.$style.isLoading]: this.isLoading,
        [this.$style.inputPadLef]: this.prefixIcon,
        [this.$style.inputPadRig]: this.suffixIcon,
      }]);
    },
  },

  methods: {
    createEvent(e) {
      return {
        name: this.model || this.name,
        value: e.currentTarget.value,
        event: e,
      };
    },

    onKeydown(e) {
      this.$emit('keydown', this.createEvent(e));
    },

    onChange(e) {
      this.$emit('change', this.createEvent(e));
    },

    onBlur(e) {
      this.$emit('blur', this.createEvent(e));
    },

    onFocus(e) {
      this.$emit('focus', this.createEvent(e));
    },
  },
};
</script>

<style module>
  .container {
    position: relative;
  }

  .input {
    width: 100%;
    border: 1px solid #DADADA;
    border-radius: 3px;
    color: #333333;
    -webkit-appearance: none;
  }

  .input:-moz-ui-invalid {
    box-shadow: none;
  }

  .small {
    height: 36px;
    padding: 11px 15px;
    font-size: 14px;
  }

  .medium {
    height: 45px;
    padding: 11px 15px;
    font-size: 15px;
  }

  .inputPadLef {
    padding-left: 35px;
  }

  .inputPadRig {
    padding-right: 35px;
  }

  .hasError {
    border-color: #D30B3F;
  }

  .isLoading {
    padding-right: 50px;
  }

  .input:focus {
    border-width: 2px;
    border-color: #00AAE5;
  }

  .input:disabled {
    background-color: #F9F9F9;
    border-color: #DADADA;
  }

  .input::placeholder {
    color: #AAAAAA;
    opacity: 1;
  }

  .loader {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 18px;
    animation: rotate-loader 1.5s infinite;
  }

  .inputIcon {
    position: absolute;
    top: 50%;
    font-size: 20px;
    transform: translateY(-50%);
  }

  .prefixIcon {
    left: 10px;
  }

  .suffixIcon {
    right: 10px;
  }

  @keyframes rotate-loader {
    0% {
      transform: translateY(-50%) rotate(0deg);
    }

    100% {
      transform: translateY(-50%) rotate(360deg);
    }
  }
</style>
