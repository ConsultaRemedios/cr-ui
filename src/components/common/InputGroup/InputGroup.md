```
<InputGroup>
  <BaseInput />
  <BaseButton >Verificar</BaseButton>
</InputGroup>
```