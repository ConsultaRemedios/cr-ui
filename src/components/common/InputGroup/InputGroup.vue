<script>
export default {
  name: 'InputGroup',
  render(createElement) {
    this.$slots.default.forEach((item) => {
      const vnode = item;

      if (/BaseInput/.test(vnode.tag)) {
        vnode.data.class = this.$style.customInput;
      }
      if (/BaseButton/.test(vnode.tag)) {
        vnode.data.class = this.$style.customButton;
      }
    });

    return createElement(
      'div',
      {
        attrs: {
          class: this.$style.wrapper,
        },
      },
      this.$slots.default,
    );
  },
};
</script>

<style module>
  .wrapper {
    display: flex;
  }

  .customInput {
    width: 100%;
  }

  .customInput input {
    border-bottom-right-radius: 0px;
    border-top-right-radius: 0px;
  }

  .customInput input:focus {
    border: 1px solid #DADADA;
  }

  .customButton {
    border-bottom-left-radius: 0px;
    border-top-left-radius: 0px;
  }
</style>
