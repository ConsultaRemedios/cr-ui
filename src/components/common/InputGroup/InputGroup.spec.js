import { mount } from '@vue/test-utils';
import InputGroup from './InputGroup.vue';
import BaseButton from './../BaseButton';
import BaseInput from './../BaseInput';

describe('InputGroup', () => {
  it('matches the snapshot', () => {
    const wrapper = mount(InputGroup, {
      slots: {
        default: [
          BaseButton,
          BaseInput,
        ]
      }
    });

    const input = wrapper.find(BaseInput);
    const button = wrapper.find(BaseButton);

    expect(input.classes()).toContain('customInput');
    expect(button.classes()).toContain('customButton');
    expect(wrapper.element).toMatchSnapshot();
  });
})
