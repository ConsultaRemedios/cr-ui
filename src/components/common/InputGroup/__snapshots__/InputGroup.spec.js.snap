// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroup matches the snapshot 1`] = `
<div
  class="wrapper"
>
  <button
    class="button medium success buttonShadow customButton"
  >
    <!---->
     
    <!---->
     
    <span
      class=""
    />
  </button>
  <div
    class="container customInput"
  >
    <!---->
     
    <input
      class="medium input"
      name=""
    />
     
    <!---->
     
    <!---->
  </div>
</div>
`;
